superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,440 @@
+    return _TrainingConfiguration(identifier=identifier, kwargs=kwargs)
+
+
+@dc.dataclass
+class LLM(Model):
+    """
+    LLM model based on `transformers` library.
+    Parameters:
+    : param identifier: model identifier
+    : param model_name_or_path: model name or path
+    : param bits: quantization bits, [4, 8], default is None
+    : param model_kwargs: model kwargs,
+        all the kwargs will pass to `transformers.AutoModelForCausalLM.from_pretrained`
+    : param tokenizer_kwags: tokenizer kwargs,
+        all the kwargs will pass to `transformers.AutoTokenizer.from_pretrained`
+    """
+
+    identifier: str = ""
+    model_name_or_path: str = "facebook/opt-125m"
+    bits: t.Optional[int] = None
+    object: t.Optional[transformers.Trainer] = None
+    model_kwargs: t.Dict = dc.field(default_factory=dict)
+    tokenizer_kwags: t.Dict = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        self.identifier = self.identifier or self.model_name_or_path
+        # overwrite model kwargs
+        if self.bits is not None:
+            self.model_kwargs["load_in_4bit"] = self.bits == 4
+            self.model_kwargs["load_in_8bit"] = self.bits == 8
+        super().__post_init__()
+
+    def init_model_and_tokenizer(self):
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_name_or_path,
+            **self.model_kwargs,
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name_or_path,
+            **self.tokenizer_kwags,
+        )
+        tokenizer.pad_token = tokenizer.pad_token or tokenizer.unk_token
+        return model, tokenizer
+
+    def create_trainer(
+        self, train_dataset, eval_dataset, training_args, **kwargs
+    ) -> transformers.Trainer:
+        trainer = transformers.Trainer(
+            model=self.model,
+            tokenizer=self.tokenizer,
+            args=training_args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            **kwargs,
+        )
+        return trainer
+
+    def init(self):
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+    def _fit(
+        self,
+        X: t.Any,
+        y: t.Optional[t.Any] = None,
+        configuration: t.Optional[_TrainingConfiguration] = None,
+        data_prefetch: bool = False,
+        db: t.Optional["Datalayer"] = None,
+        metrics: t.Optional[t.Sequence["Metric"]] = None,
+        select: t.Optional["Select"] = None,
+        validation_sets: t.Optional[t.Sequence[t.Union[str, "_Dataset"]]] = None,
+        **kwargs,
+    ):
+        assert configuration is not None, "configuration must be provided"
+
+        training_args = LLMTrainingArguments(**configuration.kwargs)  # type: ignore
+
+        # get device map
+        device_map: t.Union[None, str, t.Dict[str, int]] = None
+        if os.environ.get("LOCAL_RANK") is not None:
+            device_map = {"": int(os.environ.get("LOCAL_RANK", "0"))}
```

Comment:
we have a device on loading in pre or post create in `Model`, please check if that do not clash with this

Reply:
sorry its under def on_load

Reply:
He is only used for training and will not affect the other side.  
In addition, during our subsequent training process, this device_map needs to be calculated dynamically.